File: 2022-2023/sourcecode/lab/lab11/FoodyJava/app/src/main/java/pl/udu/uwr/pum/foodyjava/ui/fragments/FavoriteFragment.java
package pl.udu.uwr.pum.foodyjava.ui.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import pl.udu.uwr.pum.foodyjava.R;
import pl.udu.uwr.pum.foodyjava.adapters.MealComparator;
import pl.udu.uwr.pum.foodyjava.adapters.favorite.FavoriteAdapter;
import pl.udu.uwr.pum.foodyjava.adapters.meallist.MealAdapter;
import pl.udu.uwr.pum.foodyjava.databinding.FragmentFavoriteBinding;
import pl.udu.uwr.pum.foodyjava.databinding.FragmentMealDetailBinding;
import pl.udu.uwr.pum.foodyjava.ui.MealViewModel;


public class FavoriteFragment extends Fragment {

    private FragmentFavoriteBinding binding;
    private MealViewModel viewModel;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentFavoriteBinding.inflate(inflater, container, false);
        viewModel = new ViewModelProvider(requireActivity()).get(MealViewModel.class);
        viewModel.getAllMeals();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        FavoriteAdapter adapter = new FavoriteAdapter(new MealComparator());
        setupRecyclerView(adapter);

        viewModel.getFavorites().observe(getViewLifecycleOwner(), adapter::submitList);

        swipeToDelete(adapter);
    }

    private void setupRecyclerView(FavoriteAdapter adapter){
        binding.favoriteRV.setAdapter(adapter);
        binding.favoriteRV.setLayoutManager(new LinearLayoutManager(requireContext()));
    }

    private void swipeToDelete(FavoriteAdapter adapter) {
        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0,
                ItemTouchHelper.RIGHT | ItemTouchHelper.LEFT) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                viewModel.delete(adapter.getItemAt(viewHolder.getAdapterPosition()));
            }
        }).attachToRecyclerView(binding.favoriteRV);
    }
}

File: 2022-2023/sourcecode/lab/lab11/Foody/app/src/main/java/pl/udu/uwr/pum/foody/ui/favorite/FavoriteFragment.kt
package pl.udu.uwr.pum.foody.ui.favorite

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import pl.udu.uwr.pum.foody.adapters.FoodAdapter
import pl.udu.uwr.pum.foody.adapters.FoodComparator
import pl.udu.uwr.pum.foody.databinding.FragmentFavoriteBinding
import pl.udu.uwr.pum.foody.ui.FoodViewModel

class FavoriteFragment : Fragment() {

    private lateinit var binding: FragmentFavoriteBinding

    private val foodViewModel: FoodViewModel by activityViewModels()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentFavoriteBinding.inflate(layoutInflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val adapter = FoodAdapter(FoodComparator())
        setupRecyclerView(adapter)
        foodViewModel.readAllData.observe(viewLifecycleOwner, adapter::submitList)
        swipeToDelete(adapter)
    }

    private fun setupRecyclerView(foodAdapter: FoodAdapter) {
        binding.favoriteRV.apply {
            adapter = foodAdapter
            layoutManager = LinearLayoutManager(requireContext())
        }
    }

    private fun swipeToDelete(adapter: FoodAdapter) {
        ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(
            0,
            ItemTouchHelper.RIGHT or ItemTouchHelper.LEFT
        ) {
            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                foodViewModel.delete(adapter.getItemAt(viewHolder.adapterPosition))
            }
        }).attachToRecyclerView(binding.favoriteRV)
    }
}